whistleblower/whistleblower_bot/process_event.py

```diff
@@ -0,0 +1,191 @@
+    else:
+        if "purge config" in message:
+            _purge_user_config(event)
+        elif "get config" in message:
+            _get_user_config(event)
+        else:
+            _chat_post_message(event.get("channel"), utils.HELP_MESSAGE)
+
+
+def _get_user_config(event):
+    user = event.get("user")
+    if not user:
+        logger.warning("User is not present in incomming event, could not retrieve config")
+        return
+    if utils.redis_client.exists(user):
+        user_conf = json.loads(utils.redis_client.get(user))
+        _ack_config_get_msg(user_conf["phrases"], user_conf["channels"], event.get("channel"))
+    else:
+        _chat_post_message(event.get("channel"), message="You don't have any configuration!")
+
+
+def _ack_config_get_msg(phrases, channels, chan_post, base_msg=None):
+    msg = utils.DISPLAY_CONF_MESSAGE.format(
+        "\n".join(phrase for phrase in phrases), " ".join(channels)
+    )
+    if base_msg:
+        msg = base_msg + msg
+    _chat_post_message(chan_post, msg)
+
+
+def _purge_user_config(event):
+    user_id = event.get("user")
+
+    if utils.redis_client.exists(user_id):
+        utils.redis_client.delete(user_id)
+
+    for chan_id in utils.redis_client.scan_iter(match="C*"):
+        chan_conf = json.loads(utils.redis_client.get(chan_id))
+        phrases_users = chan_conf.get("phrases", {})
+
+        stale_phrases = []
+        for phrase, users in phrases_users.items():
+            if user_id in users:
+                users.remove(user_id)
+                if not users:
+                    stale_phrases.append(phrase)
+        for phrase in stale_phrases:
+            del phrases_users[phrase]
+
+        utils.redis_client.set(chan_id, json.dumps(chan_conf))
+
+    _chat_post_message(channel=event.get("channel"), message="I have purged your configuration!")
+
+
+def _update_user_config(event, phrases, channels):
+    user = event.get("user")
+    if not user:
+        logger.error("User is not present in the message description, cannot set the config")
+        return
+
+    user_conf = {"phrases": phrases, "channels": channels}
+    utils.redis_client.set(user, json.dumps(user_conf))
+
+    # stip phrases of quotes representation symbols
+    chan_phrases = [phrase.strip(utils.QUOTE_PREFIX) for phrase in phrases]
+
+    for chan in channels:
+        chan_id = chan.split("|")[0].strip("<").strip("#")
+
+        if not utils.redis_client.exists(chan_id):
+            chan_conf = {
+                "long_id": chan,
+                "phrases": {phrase: [user] for phrase in chan_phrases},
+            }
+        else:
+            chan_conf = json.loads(utils.redis_client.get(chan_id))
+            for phrase in chan_phrases:
+                phrase_users = chan_conf["phrases"].setdefault(phrase, [])
+                if user not in phrase_users:
+                    phrase_users.append(user)
+        utils.redis_client.set(chan_id, json.dumps(chan_conf))
+
```

Comment:
It seems like the config structure in Redis is complex, which brings additional complexity for setting/purging user configs. How would you redesign it to optimize and simplify the config management?

Reply:
@mmatviyiv let me provide some context on what i wanted to achieve while designing the config processing sub-routines. 

Actually the redis db layout is pretty simple, there will be only string data items with the number of keys = m + n, where m is the number of unique registered (added configuration) users and n - number of unique channels in which at least one user is listening.

The configuration data structure also matters, since several users may have defined same phrases to listen to and a several channels may be monitored for a particular phrase, so there is no clear 1:1 relation in either phrase to user or channel to phrase, the only unique identifier would consists of all three items - user+phrase+channel. Thus there is no means to produce small effective schema for such kind of hierarchy in redis, one need to have several sets or smth, but then it requires several request to retrieve the relevant data on `message.channels` processing step.


The main goal of mine was reduce the number of redis queries from `message.channels` event processing function - `process_chat_message`, to as small as possible, and which currently is exactly two in the worst case scenario - checking for the key and retrieving the key's value if it exists. 
I have done so, since this piece of code will be called most frequently as it has to analyze events for every message sent in every public channel the bot is added to, on the other hand the frequency of config update is relatively low in comparison, so i have decided to sacrifice performance on the python side

Reply:
wow, thanks a lot for such a detailed response. That makes sense.
Redis provides one trick that can simplify your code, improve performance and the most important - deal with concurrency issues. Lua scripting - which runs like one single atomic command. Really cool feature.

Reply:
@mmatviyiv thanks, will look into lua scripting feature, should i still do the refactoring of the config update/purge using it in this PR?

Reply:
no updates needed, thanks, it was more like a hypothetical question